File: tvn-cosine.ai/tvn-cosine.ai/Search/csp/TreeCspSolver.cs
﻿using System;
using System.Collections.Generic;
using System.Threading;
using tvn.cosine.ai.util;

namespace tvn.cosine.ai.search.csp
{
    /**
     * Artificial Intelligence A Modern Approach (3rd Ed.): Figure 6.11, Page
     * 224.<br>
     * <br>
     * <p>
     * <pre>
     * <code>
     * function TREE-CSP-SOLVER(csp) returns a solution, or failure
     * 		inputs: csp, a CSP with components X, D, C
     * 		n &larr; number of variables in X
     * 		assignment &larr; an empty assignment
     * 		root  &larr; any variable in X
     * 		X &larr; TOPOLOGICALSORT(X, root )
     * 		for j = n down to 2 do
     * 			MAKE-ARC-CONSISTENT(PARENT(Xj),Xj )
     * 			if it cannot be made consistent then return failure
     * 		for i = 1 to n do
     * 			assignment[Xi] &larr; any consistent value from Di
     * 			if there is no consistent value then return failure (*)
     * 		return assignment
     * </code>
     *
     * <pre>
     *
     * Figure 6.11 The TREE-CSP-SOLVER algorithm for solving tree-structured CSPs.
     * If the CSP has a solution, we will find it in linear time; if not, we will
     * detect a contradiction. Comment to (*) (RL): If no empty domain was found in the
     * previous loop, this will only happen if n == 1.
     *
     * @author Ruediger Lunde
     * @author Anurag Rai
     */
    public class TreeCspSolver<VAR, VAL> : CspSolver<VAR, VAL>
        where VAR : Variable
    {
        private bool useRandom;

        public TreeCspSolver<VAR, VAL> UseRandom(bool b)
        {
            useRandom = b;
            return this;
        }

        public override Assignment<VAR, VAL> solve(CSP<VAR, VAL> csp, CancellationToken cancellationToken)
        {
            Assignment<VAR, VAL> assignment = new Assignment<VAR, VAL>();
            // Select a root from the List of Variables
            VAR root = useRandom ? Util.selectRandomlyFromList(csp.getVariables()) : csp.getVariables()[0];
            // Sort the variables in topological order
            List<VAR> orderedVars = new List<VAR>();
            IDictionary<VAR, Constraint<VAR, VAL>> parentConstraints = new Dictionary<VAR, Constraint<VAR, VAL>>();
            topologicalSort(csp, root, orderedVars, parentConstraints);
            if (csp.getDomain(root).isEmpty())
                return null; // CSP has no solution! (needed if orderedVars.size() == 1)

            // Establish arc consistency from top to bottom (starting at the bottom).
            csp = csp.copyDomains(); // do not change the original CSP!
            for (int i = orderedVars.Count - 1; i > 0; i--)
            {
                VAR var = orderedVars[i];
                Constraint<VAR, VAL> constraint = parentConstraints[var];
                VAR parent = csp.getNeighbor(var, constraint);
                if (makeArcConsistent(parent, var, constraint, csp))
                {
                    fireStateChanged(csp, null, parent);
                    if (csp.getDomain(parent).isEmpty())
                        return null; // CSP has no solution!
                }
            }

            // Assign values to variables from top to bottom.
            for (int i = 0; i < orderedVars.Count; i++)
            {
                VAR var = orderedVars[i];
                foreach (VAL value in csp.getDomain(var))
                {
                    assignment.add(var, value);
                    if (assignment.isConsistent(csp.getConstraints(var)))
                    {
                        fireStateChanged(csp, assignment, var);
                        break;
                    }
                }
            }

            return assignment;
        }

        /**
         * Computes an explicit representation of the tree structure and a total order which is consistent with the
         * parent-child relations. If the provided CSP has not the required properties (CSP contains only binary
         * constraints, constraint graph is tree-structured and connected), an exception is thrown.
         *
         * @param csp               A CSP
         * @param root              A root variable
         * @param orderedVars       The computed total order (initially empty)
         * @param parentConstraints The tree structure, maps a variable to the constraint representing the arc to the parent
         *                          variable (initially empty)
         */
        private void topologicalSort(CSP<VAR, VAL> csp, VAR root, List<VAR> orderedVars,
                                     IDictionary<VAR, Constraint<VAR, VAL>> parentConstraints)
        {
            orderedVars.Add(root);
            parentConstraints.Add(root, null);
            int currParentIdx = -1;
            while (currParentIdx < orderedVars.Count - 1)
            {
                currParentIdx++;
                VAR currParent = orderedVars[currParentIdx];
                int arcsPointingUpwards = 0;
                foreach (Constraint<VAR, VAL> constraint in csp.getConstraints(currParent))
                {
                    VAR neighbor = csp.getNeighbor(currParent, constraint);
                    if (neighbor == null)
                        throw new ArgumentException("Constraint " + constraint + " is not binary.");
                    if (parentConstraints.ContainsKey(neighbor))
                    { // faster than orderedVars.contains(neighbor)!
                        arcsPointingUpwards++;
                        if (arcsPointingUpwards > 1)
                            throw new ArgumentException("CSP is not tree-structured.");
                    }
                    else
                    {
                        orderedVars.Add(neighbor);
                        parentConstraints.Add(neighbor, constraint);
                    }
                }
            }
            if (orderedVars.Count < csp.getVariables().Count)
                throw new ArgumentException("Constraint graph is not connected.");
        }

        /**
         * Establishes arc-consistency for (xi, xj).
         * @return value true if the domain of xi was reduced.
         */
        private bool makeArcConsistent(VAR xi, VAR xj, Constraint<VAR, VAL> constraint, CSP<VAR, VAL> csp)
        {
            Domain<VAL> currDomain = csp.getDomain(xi);
            List<VAL> newValues = new List<VAL>(currDomain.size());
            Assignment<VAR, VAL> assignment = new Assignment<VAR, VAL>();
            foreach (VAL vi in currDomain)
            {
                assignment.add(xi, vi);
                foreach (VAL vj in csp.getDomain(xj))
                {
                    assignment.add(xj, vj);
                    if (constraint.isSatisfiedWith(assignment))
                    {
                        newValues.Add(vi);
                        break;
                    }
                }
            }
            if (newValues.Count < currDomain.size())
            {
                csp.setDomain(xi, new Domain<VAL>(newValues));
                return true;
            }
            return false;
        }
    }

}

File: tvn-cosine.ai/tvn-cosine.ai.v2/search/basic/csp/TreeCSPSolver.java
namespace aima.core.search.basic.csp;

using java.util.ArrayList;
using java.util.HashMap;
using java.util.LinkedList;
using java.util.List;
using java.util.Map;
using java.util.Queue;

using aima.core.search.api.Assignment;
using aima.core.search.api.CSP;
using aima.core.search.api.Constraint;
using aima.core.search.api.SearchForAssignmentFunction;
using aima.core.search.basic.support.BasicAssignment;

/**
 * 
 * Artificial Intelligence A Modern Approach (4th Ed.): Figure ??, Page ??.<br>
 * <br>
 * 
 * <pre>
 * <code>
 * function TREE-CSP-SOLVER(csp) returns a solution, or failure
 *    inputs: csp, a CSP with components X, D, C
 * 
 *    n &larr; number of variables in X
 *    assignment &larr; an empty assignment
 *    root  &larr; any variable in X
 *    X &larr; TOPOLOGICALSORT(csp, root)
 *    for j = n down to 2 do
 *       MAKE-ARC-CONSISTENT(csp, PARENT(X<sub>j</sub>), X<sub>j</sub>)
 *       if it cannot be made consistent then return failure
 *    for i = 1 to n do
 *       assignment[X<sub>i</sub>] &larr; any consistent value from D<sub>i</sub>
 *       if there is no consistent value then return failure
 *    return assignment
 * </code>
 * 
 * <pre>
 * 
 * Figure ?? The TREE-CSP-SOLVER algorithm for solving tree-structured CSPs. If
 * the CSP has a solution, we will find it in linear time; if not, we will
 * detect a contradiction.
 * 
 * @author Ciaran O'Reilly
 * @author Anurag Rai
 * 
 */
public class TreeCSPSolver implements SearchForAssignmentFunction {

	// function TREE-CSP-SOLVER(csp) returns a solution, or failure
	// inputs: csp, a CSP with components X, D, C
	 
	public Assignment apply(CSP csp) {
		// n <- number of variables in X
		int n = csp.getVariables().size();
		// assignment <- an empty assignment
		Assignment assignment = newEmptyAssignment();
		// root <- any variable in X
		String root = anyVariable(csp);
		// X <- TOPOLOGICALSORT(csp, root)
		TopologicalSort X = topologicalSort(csp, root);
		// for j = n down to 2 do
		for (int j = n - 1; j >= 1; j--) {
			// MAKE-ARC-CONSISTENT(csp, PARENT(X<sub>j</sub>), X<sub>j</sub>)
			makeArcConsistent(csp, X.parent(X.variables.get(j)), X.variables.get(j));
			// if it cannot be made consistent then return failure
			if (csp.isInconsistent()) {
				return failure();
			}
		}
		// for i = 1 to n do
		for (int i = 0; i < n; i++) {
			// assignment[X<sub>i</sub>] <- any consistent value from
			// D<sub>i</sub>
			String Xi = X.variables.get(i);
			for (Object value : csp.getDomain(Xi).getValues()) {
				assignment.add(Xi, value);
				if (assignment.isConsistent(csp)) {
					break;
				}
			}
			// if there is no consistent value then return failure
			if (!assignment.isConsistent(csp)) {
				return failure();
			}
		}

		return assignment;
	}

	//
	// Supporting Code
	public static class TopologicalSort {
		public List<String> variables = new ArrayList<>();
		public IDictionary<String, String> parents = new HashMap<>();

		public String parent(String childVariable) {
			return parents.get(childVariable);
		}

		 
		public override string ToString() {
			return "TopologicalSort[sort=" + variables + ", parents=" + parents + "]";
		}
	}

	public Assignment newEmptyAssignment() {
		return new BasicAssignment();
	}

	public String anyVariable(CSP csp) {
		return csp.getVariables().get(0);
	}

	public TopologicalSort topologicalSort(CSP csp, String root) {
		if (!csp.isTree()) {
			throw new IllegalArgumentException("CSP is not structured as a tree");
		}
		TopologicalSort result = new TopologicalSort();
		result.variables.add(root);

		Queue<String> toProcess = new LinkedList<>();
		toProcess.add(root);
		while (toProcess.size() > 0) {
			String current = toProcess.remove();
			for (Constraint c : csp.getNeighboringConstraints(current)) {
				for (String scopeVar : c.getScope()) {
					if (!scopeVar.Equals(current)) {
						if (!result.variables.contains(scopeVar)) {
							result.variables.add(scopeVar);
							result.parents.put(scopeVar, current);
							toProcess.add(scopeVar);
						}
					}
				}
			}
		}

		// Just in case!
		if (result.variables.size() != csp.getVariables().size()) {
			throw new IllegalStateException("Unable to perform a topological sort correctly, computed=" + result);
		}

		return result;
	}

	public void makeArcConsistent(CSP csp, String parentVar, String childVar) {
		csp.getNeighboringConstraints(parentVar).stream().filter(constraint -> constraint.getScope().contains(childVar))
				.forEach(constraint -> {
					for (Object parentValue : csp.getDomain(parentVar).getValues()) {
						if (!csp.getDomain(childVar).getValues().stream().anyMatch(childValue -> {
							Object[] values = new Object[2];
							values[constraint.getScope().indexOf(parentVar)] = parentValue;
							values[constraint.getScope().indexOf(childVar)] = childValue;
							return constraint.getRelation().isMember(values);
						})) {
							csp.getDomain(parentVar).delete(parentValue);
						}
					}
				});
	}

	public Assignment failure() {
		return null;
	}
}